client/executor/src/wasm_runtime.rs

```diff
@@ -213,83 +214,90 @@ impl RuntimeCache {
 		allow_missing_func_imports: bool,
 		f: F,
 	) -> Result<Result<R, Error>, Error>
-		where F: FnOnce(
+		where F: Fn(
 			&Arc<dyn WasmModule>,
 			&dyn WasmInstance,
 			Option<&RuntimeVersion>,
 			&mut dyn Externalities)
 		-> Result<R, Error>,
 	{
 		let code_hash = &runtime_code.hash;
-		let heap_pages = runtime_code.heap_pages.unwrap_or(default_heap_pages);
-
-		let mut runtimes = self.runtimes.lock(); // this must be released prior to calling f
-		let pos = runtimes.iter().position(|r| r.as_ref().map_or(
-			false,
-			|r| r.wasm_method == wasm_method &&
-				r.code_hash == *code_hash &&
-				r.heap_pages == heap_pages
-		));
-
-		let runtime = match pos {
-			Some(n) => runtimes[n]
-				.clone()
-				.expect("`position` only returns `Some` for entries that are `Some`"),
-			None =>  {
-				let code = runtime_code.fetch_runtime_code().ok_or(WasmError::CodeNotFound)?;
-
-				#[cfg(not(target_os = "unknown"))]
-				let time = std::time::Instant::now();
-
-				let result = create_versioned_wasm_runtime(
-					&code,
-					code_hash.clone(),
-					ext,
-					wasm_method,
-					heap_pages,
-					host_functions.into(),
-					allow_missing_func_imports,
-					self.max_runtime_instances,
-					self.cache_path.as_deref(),
-				);
+		let mut heap_pages = default_heap_pages;
+		loop {
+			let try_heap_pages = runtime_code.heap_pages.clone().unwrap_or(heap_pages);
```

Comment:
looks to me like this will skip the growing `heap_pages` variable each loop if `runtime_code.heap_pages` is `Some`.